### PR TITLE
Optional wildcards

### LINKED ISSUE #24: Feature request: optional wildcards
It would be nice to be able to match 0 or more characters for a wildcard:

`/books/:genre/*?` => match both `/books/abc` and `/books/abc/xyz`

### PR DESCRIPTION
**Pull request by benmerckx**
Closes #24

### CODE REVIEW

**Comment by molefrog** [comment]
Hi @lukeed, just wanted to ask to take a brief look at this. 
This is the only thing that blocks us from switching to regexparam in https://github.com/molefrog/wouter.

File: src/index.js [hunk 1]
```diff
@@ -6,8 +6,9 @@ export function parse(str, loose) {
 	while (tmp = arr.shift()) {
 		c = tmp[0];
 		if (c === '*') {
+			o = tmp[1] === '?';
 			keys.push('wild');
-			pattern += '/(.*)';
+			pattern += o ? '(?:/(.*))?' : '/(.*)';
```
Comment (lukeed) [inline src/index.js:11]:
```suggestion
			pattern += tmp[1] === '?' ? '(?:/(.*))?' : '/(.*)';
```

File: src/index.js [hunk 2]
```diff
@@ -6,8 +6,9 @@ export function parse(str, loose) {
 	while (tmp = arr.shift()) {
 		c = tmp[0];
 		if (c === '*') {
+			o = tmp[1] === '?';
```
Comment (lukeed) [inline src/index.js:9]:
```suggestion
```

**Review by lukeed — CHANGES_REQUESTED** [review]

**Review by lukeed — APPROVED** [review]
Thank you for this :) It's a breaking change due to how the `"/books/*?"` pattern will now match `"/books"` (shown in test), but I think it's a change worth having.

Cutting this as a 3.0 release

**Comment by molefrog** [comment]
@lukeed Thank you! We added it back as a dependency and it works well. https://github.com/molefrog/wouter/pull/395

**Pull request merged**